src/adapton2.mo

```diff
@@ -0,0 +1,638 @@
+  public type LogBufStack<Name, Val, Error, Env, Exp> = L.List<LogEventBuf<Name, Val, Error, Env, Exp>>;
+
+
+  public type Context<Name, Val, Error, Env, Exp> = {
+    var agent: {#editor; #archivist};
+    var edges: EdgeBuf<Name, Val, Error, Env, Exp>;
+    var stack: Stack<Name>;
+    var store: Store<Name, Val, Error, Env, Exp>;
+    // logging for debugging; not essential for other state:
+    var logFlag: Bool;
+    var logBuf: LogEventBuf<Name, Val, Error, Env, Exp>;
+    var logStack: LogBufStack<Name, Val, Error, Env, Exp>;
+    var eval: E.Eval<Name, Val, Error, Env, Exp>;
+  };
+
+  // class accepts the associated operations over the 5 user-defined type params
+  public class Engine<_Name, _Val, _Error, _Env, _Exp>(Eval:E.Eval<_Name, _Val, _Error, _Env, _Exp>) {
+    // to do: define the public interface of Adapton using Eval to implement expression evaluation ...
+    // ... by following original cleanSheets implementation (mostly copying it, now commented out below).
+
+/*
+    public func closureEq(c1:Closure, c2:Closure) : Bool {
+      Eval.envEq(c1.env, c2.env)
+      and Eval.expEq(c1.exp, c2.exp)
+    };
+
+    public func logEventsEq (e1:[LogEvent], e2:[LogEvent]) : Bool {
+      if (e1.len() == e2.len()) {
+        for (i in e1.keys()) {
+          if (logEventEq(e1[i], e2[i])) {
+            /* continue */
+          } else {
+            return false
+          }
+        };
+        true
+      } else { false }
+    };
+
+    public func logEventEq (e1:LogEvent, e2:LogEvent) : Bool {
+      switch (e1, e2) {
+      case (#put(n1, v1, es1), #put(n2, v2, es2)) {
+             Eval.nameEq(n1, n2) and Eval.valEq(v1, v2) and logEventsEq(es1, es2)
+           };
+      case (#putThunk(n1, c1, es1), #putThunk(n2, c2, es2)) {
+             P.nyi()
+           };
+      case (#get(n1, r1, es1), #get(n2, r2, es2)) {
+             Eval.nameEq(n1, n2) and Eval.resultEq(r1, r2) and logEventsEq(es1, es2)
+           };
+      case (#dirtyIncomingTo(n1, es1), #dirtyIncomingTo(n2, es2)) {
+             Eval.nameEq(n1, n2) and logEventsEq(es1, es2)
+           };
+      case (#dirtyEdgeFrom(n1, es1), #dirtyEdgeFrom(n2, es2)) {
+             Eval.nameEq(n1, n2) and logEventsEq(es1, es2)
+           };
+      case (#cleanEdgeTo(n1, f1, es1), #cleanEdgeTo(n2, f2, es2)) {
+             Eval.nameEq(n1, n2) and f1 == f2 and logEventsEq(es1, es2)
+           };
+      case (#cleanThunk(n1, f1, es1), #cleanThunk(n2, f2, es2)) {
+             Eval.nameEq(n1, n2) and f1 == f2 and logEventsEq(es1, es2)
+           };
+      case (#evalThunk(n1, r1, es1), #evalThunk(n2, r2, es2)) {
+             Eval.nameEq(n1, n2) and Eval.resultEq(r1, r2) and logEventsEq(es1, es2)
+           };
+      case (_, _) {
+             false
+           }
+      }
+
+    };
+
+
+    public func init(_logFlag:Bool) : Context {
+      // to do -- compiler bug? -- IR typing issue when this line is inlined to its use below:
```

Comment:
is that bug still there?

Reply:
Nope, this must have been from an old compiler run, by mistake.